include/highfive/h5easy_bits/H5Easy_Eigen.hpp

```diff
@@ -19,91 +19,86 @@ namespace H5Easy {
-            std::decay<T>::type::MaxColsAtCompileTime>,
-        0,
-        Eigen::InnerStride<1>> row_major(data);
-
-    dataset.write_raw(row_major.data());
-}
-
-// create DataSet and write data
-template <class T>
-static DataSet dump_impl(File& file, const std::string& path, const T& data) {
-    using value_type = typename std::decay<T>::type::Scalar;
-    detail::createGroupsToDataSet(file, path);
-    DataSet dataset = file.createDataSet<value_type>(path, DataSpace(shape(data)));
-    detail::eigen::write(dataset, data);
-    file.flush();
-    return dataset;
-}
-
-// replace data of an existing DataSet of the correct size
-template <class T>
-static DataSet overwrite_impl(File& file, const std::string& path, const T& data) {
-    DataSet dataset = file.getDataSet(path);
-    if (dataset.getDimensions() != shape(data)) {
-        throw detail::error(file, path, "H5Easy::dump: Inconsistent dimensions");
-    }
-    detail::eigen::write(dataset, data);
-    file.flush();
-    return dataset;
-}
-
-// load from DataSet
-// convert to ColMajor if needed (HDF5 always stores row-major)
-template <class T>
-struct load_impl {
-    static T run(const File& file, const std::string& path) {
+template<typename T>
+struct io_impl<T,typename std::enable_if<std::is_base_of<Eigen::DenseBase<T>,T>::value>::type>: public io_impl_base<T> {
+		
+	// return the shape of Eigen::DenseBase<T> object as size 1 or 2 "std::vector<size_t>"
+	inline static std::vector<size_t> shape(const T& data) {
+		if (std::decay<T>::type::RowsAtCompileTime == 1) {
+			return {static_cast<size_t>(data.cols())};
+		}
+		if (std::decay<T>::type::ColsAtCompileTime == 1) {
+			return {static_cast<size_t>(data.rows())};
+		}
+		return {static_cast<size_t>(data.rows()),
+				static_cast<size_t>(data.cols())};
+	}
+	
+	using EigenIndex = Eigen::DenseIndex;
+
+	// get the shape of a "DataSet" as size 2 "std::vector<Eigen::Index>"
+	inline static std::vector<EigenIndex> shape(const File& file,
+                                               const std::string& path,
+                                               const DataSet& dataset,
+                                               int RowsAtCompileTime) {
+		std::vector<size_t> dims = dataset.getDimensions();
+
+		if (dims.size() == 1 && RowsAtCompileTime == 1) {
+			return std::vector<EigenIndex>{1u, static_cast<EigenIndex>(dims[0])};
+		}
+		if (dims.size() == 1) {
+			return std::vector<EigenIndex>{static_cast<EigenIndex>(dims[0]), 1u};
+		}
+		if (dims.size() == 2) {
+			return std::vector<EigenIndex>{static_cast<EigenIndex>(dims[0]),
+                                           static_cast<EigenIndex>(dims[1])};
+		}
+
+		throw detail::error(file, path, "H5Easy::load: Inconsistent rank");
+	}
+
+	// write to open DataSet of the correct size
+	// (use Eigen::Ref to convert to RowMajor; no action if no conversion is needed)
+	inline static void write(DataSet& dataset, const T& data) {
+		Eigen::Ref<
+			const Eigen::Array<
```

Comment:
Just out or curiosity, what is the difference between `Array` and `Matrix`? I was under the impression that `Eigen` only had matrices.

Reply:
`Matrix` is... mathematical matrix (and `Vector*` is 1D matrix), Array (1D or 2D) is element-wise by default. Converting (compile-time, no allocation) between them is done with `.matrix()` and `.array()` methods. As we are using arrays just as a bunch of numbers grid, that intent is clear by using `Array`.